dex/networks/eth/contracts/ETHSwapV1.sol

```diff
@@ -0,0 +1,226 @@
+
+            swaps[k] = record;
+
+            initVal += c.value * 1 gwei;
+        }
+
+        require(initVal == msg.value, "bad val");
+    }
+
+    // isRedeemable returns whether or not a swap identified by secretHash
+    // can be redeemed using secret.
+    function isRedeemable(Contract calldata c)
+        public
+        view
+        returns (bool)
+    {
+        (, bytes32 record, uint256 blockNum) = retrieveRecord(c);
+        return blockNum != 0 && !secretValidates(record, c.secretHash);
+    }
+
+    // redeem redeems a Contract. It checks that the sender is not a contract,
+    // and that the secret hash hashes to secretHash. msg.value is tranfered
+    // from ETHSwap to the sender.
+    //
+    // To prevent reentry attack, it is very important to check the state of the
+    // contract first, and change the state before proceeding to send. That way,
+    // the nested attacking function will throw upon trying to call redeem a
+    // second time. Currently, reentry is also not possible because contracts
+    // cannot use this contract.
+    function redeem(Redemption[] calldata redemptions)
+        public
+        senderIsOrigin()
+    {
+        uint amountToRedeem = 0;
+        for (uint i = 0; i < redemptions.length; i++) {
+            Redemption calldata r = redemptions[i];
+
+            require(r.c.participant == msg.sender, "not authed");
+
+            (bytes32 k, bytes32 record, uint256 blockNum) = retrieveRecord(r.c);
+
+            // To be redeemable, the record needs to represent a valid block
+            // number.
+            require(blockNum > 0 && blockNum < block.number, "unfilled swap");
+
+            // Can't already be redeemed.
+            require(!secretValidates(record, r.c.secretHash), "already redeemed");
+
+            // Are they presenting the correct secret?
+            require(secretValidates(r.secret, r.c.secretHash), "invalid secret");
+
+            swaps[k] = r.secret;
+            amountToRedeem += r.c.value * 1 gwei;
+        }
+
+        (bool ok, ) = payable(msg.sender).call{value: amountToRedeem}("");
+        require(ok == true, "transfer failed");
+    }
+
+    // refund refunds a Contract. It checks that the sender is not a contract
+    // and that the refund time has passed. msg.value is transfered from the
+    // contract to the sender = Contract.participant.
+    //
+    // It is important to note that this also uses call.value which comes with
+    // no restrictions on gas used. See redeem for more info.
+    function refund(Contract calldata c)
+        public
+        senderIsOrigin()
+    {
+        // Is this contract even in a refundable state?
+        require(block.timestamp >= c.refundTimestamp, "locktime not expired");
+
+        // Retrieve the record.
+        (bytes32 k, bytes32 record, uint256 blockNum) = retrieveRecord(c);
+
+        // Is this swap initialized?
+        require(blockNum > 0 && blockNum <= block.number, "swap not active");
```

Comment:
I think this makes the refunded check not reachable if refunded. It also stops most checks for is redeemed. Both of which are fine, but can remove the explicit refund check I guess, or move it and the redeem check up if the error is important.